lib/vnet/daemon/service_darwin.m

```diff
@@ -0,0 +1,143 @@
+//go:build vnetdaemon
+// +build vnetdaemon
+
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "common_darwin.h"
+#include "protocol_darwin.h"
+#include "service_darwin.h"
+
+#import <Foundation/Foundation.h>
+
+#include <string.h>
+
+@interface VNEDaemonService () <NSXPCListenerDelegate, VNEDaemonProtocol>
+
+@property(nonatomic, strong, readwrite) NSXPCListener *listener;
+@property(nonatomic, readwrite) BOOL started;
+
+@property(nonatomic, readwrite) NSString *socketPath;
+@property(nonatomic, readwrite) NSString *ipv6Prefix;
+@property(nonatomic, readwrite) NSString *dnsAddr;
+@property(nonatomic, readwrite) NSString *homePath;
+@property(nonatomic, readwrite) dispatch_semaphore_t gotVnetConfigSema;
+
+@end
+
+@implementation VNEDaemonService
+
+- (id)initWithBundlePath:(NSString *)bundlePath {
+  self = [super init];
+  if (self) {
+    // Launch daemons must configure their listener with the machServiceName
+    // initializer.
+    _listener = [[NSXPCListener alloc] initWithMachServiceName:DaemonLabel(bundlePath)];
+    _listener.delegate = self;
+
+    _started = NO;
+    _gotVnetConfigSema = dispatch_semaphore_create(0);
+  }
+  return self;
+}
+
+- (void)start {
+  assert(_started == NO);
+
+  // Begin listening for incoming XPC connections.
+  [_listener resume];
+
+  _started = YES;
+}
+
+- (void)stop {
+  assert(_started == YES);
+
+  // Stop listening for incoming XPC connections.
+  [_listener suspend];
+
+  _started = NO;
+  dispatch_semaphore_signal(_gotVnetConfigSema);
+}
+
+- (void)waitForVnetConfig {
+  assert(_started == YES);
+  dispatch_semaphore_wait(_gotVnetConfigSema, DISPATCH_TIME_FOREVER);
+}
+
+#pragma mark - VNEDaemonProtocol
+
+- (void)startVnet:(VnetConfig *)vnetConfig completion:(void (^)(void))completion {
+  _socketPath = @(vnetConfig->socket_path);
+  _ipv6Prefix = @(vnetConfig->ipv6_prefix);
+  _dnsAddr = @(vnetConfig->dns_addr);
+  _homePath = @(vnetConfig->home_path);
```

Comment:
I was wondering if this code should use some kind of a mutex, perhaps [a `@synchronized` directive](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Multithreading/ThreadSafety/ThreadSafety.html#//apple_ref/doc/uid/10000057i-CH8-SW3).

There's very little information I can find about XPC and thread safety, other than [a Developer Forums thread about it](https://forums.developer.apple.com/forums/thread/70028). It seems to suggest that messages sent over XPC are serialized:

> NSXPCConnection replies are always delivered on a specific serial queue internal to the connection. There’s a comment in `<Foundation/NSXPCConnection.h>` that says as much:
>
> > Each NSXPCConnection instance has a private serial queue. This queue is used when sending messages to reply handlers, interruption handlers, and invalidation handlers.
>
> Which _thread_ runs blocks posted to that queue is not specified.

This makes the XPC part thread-safe, as in if a client makes two concurrent calls, they'll be processed serially. But I'm still calling `WaitForVnetConfig` from a separate thread which doesn't care about that serial queue at all.

I'm going to add a `@synchronized` directive here then.